specification/storagecache/resource-manager/Microsoft.StorageCache/preview/2020-08-13-aml-preview/storagecache.json

```diff
@@ -0,0 +1,2623 @@
+          },
+          "type": "array"
+        }
+      },
+      "type": "object"
+    },
+    "amlFilesystemIdentity": {
+      "description": "amlFilesystem identity properties.",
+      "properties": {
+        "principalId": {
+          "readOnly": true,
+          "type": "string",
+          "description": "The principal id of the amlFilesystem."
+        },
+        "tenantId": {
+          "readOnly": true,
+          "type": "string",
+          "description": "The tenant id associated with the amlFilesystem."
+        },
+        "type": {
+          "type": "string",
+          "description": "The type of identity used for the amlFilesystem",
+          "enum": [
+            "SystemAssigned",
+            "None"
+          ],
+          "x-ms-enum": {
+            "name": "amlFilesystemIdentityType",
+            "modelAsString": false
+          }
+        }
+      }
+    },
+    "amlFilesystemMaintenanceStatus": {
+      "description": "Properties describing the software upgrade state of the amlFilesystem.",
+      "properties": {
+        "currentFirmwareVersion": {
+          "description": "Version string of the firmware currently installed on this amlFilesystem.",
+          "type": "string",
+          "readOnly": true
+        },
+        "firmwareUpdateStatus": {
+          "description": "True if there is a firmware update ready to install on this amlFilesystem. The firmware will automatically be installed after firmwareUpdateDeadline if not triggered earlier via the upgrade operation.",
+          "enum": [
+            "available",
+            "unavailable"
+          ],
+          "x-ms-enum": {
+            "name": "FirmwareStatusType",
+            "modelAsString": true
+          },
+          "readOnly": true,
+          "type": "string"
+        },
+        "firmwareUpdateDeadline": {
+          "description": "Time at which the pending firmware update will automatically be installed on the amlFilesystem.",
+          "format": "date-time",
+          "type": "string",
+          "readOnly": true
+        },
+        "lastFirmwareUpdate": {
+          "description": "Time of the last successful firmware update.",
+          "format": "date-time",
+          "type": "string",
+          "readOnly": true
+        },
+        "pendingFirmwareVersion": {
+          "description": "When firmwareUpdateAvailable is true, this field holds the version string for the update.",
+          "type": "string",
+          "readOnly": true
+        }
+      },
+      "readOnly" : true,
+      "type": "object"
+    },
+    "amlFilesystemHealth": {
+      "description": "An indication of amlFilesystem health. Gives more information about health than just that related to provisioning.",
+      "properties": {
+        "state": {
+          "description": "List of amlFilesystem health states.",
+          "enum": [
+            "Unknown",
+            "Healthy",
+            "Down",
```

Comment:
"Healthy" also might be tricky in large clusters since one or more nodes are likely to be down at all times when dealing with hundreds of VMs.

I'll propose this set of states:  "Unknown", "Available", "Unavailable", "Degraded"

Reply:
The health state should reflect the customer's view of the filesystem, not the internal state.   So if the thing is serving files at a reasonable performance level, that's healthy as far as a customer is concerned.   Non-healthy states need to be actionable and/or have impact on the customer's ability to use the filesystem.

Reply:
Dan - I agree with that point.  I'm just wondering if "Healthy" is the best way to describe a state where the file system is generally available in the customer's eyes.

If 1 out of 200 nodes are down, then the Lustre file system is 99.5% available, so 0.5% of the accesses may be experiencing some difficulties.  In this respect, the file system is healthy unless your data is on the 0.5% of the unhealthy nodes.

I was thinking the term "Available" more directly reflects this state from the customer's perspective: the service is up, it can be mounted, and the data is generally accessible with a small number of exceptions.

Jim and I have "Degraded" in this list as well.  Defining that will require some reflection on our part, and maybe not in time for v1 until we get some experience.  For example, which percentage of VMs must be down and for how long before we declare the file system as "Degraded"?

Reply:
Should we associate a "available %" in Available state that gives a hint as to how much of the file is inaccessible?  Seems drastic to provide a "unhealthy"-like status when only one node goes down in a cluster of many 100s of nodes.

Reply:
Yeah, I was thinking of Degraded as a catch-all for "some or all of the data is accessible, but the experience is not optimal". What I'm trying to do is differentiate between Unavailable (IOW, don't expect anything to work) and "partly available". If we advertise the filesystem as Unavailable and yet reads and writes are making progress, then the customer can be confused, or can end up distrusting our health reports. The same with Healthy if they have reads stuck because an OSS is down.

Reply:
It sounds like this could work then.  "Unknown", "Available", "Unavailable", "Degraded"
Description: 
Available - The service is up and all data is accessible.
Degraded - The service is up but some data is not accessible.
Unavailable - The service is down and no data is accessible.
Unknown - The service can't be reached.  The status is unknown.

Reply:
Since the maintenanceStatus was gutted, and now just has a currentSoftwareVersion, there would be no indication that its in maintenance.  I think we at least need to add that back in here.
In addition to that transitioning was something I think cache had to note the state while it was still being created.  I'm not sure any of those states properly indicate that its being created still.

There is a provisioning state in the get but its still separate and its state is creating.

Reply:
Yeah, adding something like Maintenance or ScheduledMaintenance is a great idea.

Reply:
That's all in the provisioning state, which is dictated by ARM in most cases.  We have transitioning in health to cover these different provisioning cases.